diffkemp/snapshot.py

```diff
@@ -0,0 +1,226 @@
+        :param name: Name of the group.
+        """
+        self.fun_groups[name] = self.FunctionGroup()
+
+    def add_none_group(self):
+        """
+        Add new group with None name which is used when only a single group
+        is used.
+        """
+        self.fun_groups[None] = self.FunctionGroup()
+
+    def modules(self):
+        """
+        Get the set of all LLVM modules of all functions in the function list.
+        """
+        return set([fun.mod for group in self.fun_groups.values()
+                    for fun in group.functions.values()
+                    if fun.mod is not None])
+
+    def get_by_name(self, name, group=None):
+        """
+        Get module for the function with the given name in the given group.
+        :param name: Name of the function.
+        :param group: Group of the function.
+        """
+        return self.fun_groups[group].functions[name] \
+            if group in self.fun_groups and name in self.fun_groups[
+            group].functions \
+            else None
+
+    def filter(self, functions, group=None):
+        """
+        Filter only chosen functions in the given group.
+        :param name: List of function names to filter.
+        :param group: Group to filter.
+        """
+        self.fun_groups[group].functions = {f: self.fun_groups[
+            group].functions[f] for f in functions}
+
+    def from_yaml(self, yaml_file):
+        """
+        Load the snaphot from its YAML representation. Paths are assumed to be
+        relative to the root directory.
+        :param yaml_file: Contents of the YAML file.
+        """
+        yaml_file = yaml.safe_load(yaml_file)
+        yaml_dict = yaml_file[0]
+
+        if os.path.isdir(yaml_dict["source_kernel_dir"]):
+            self.kernel_source = KernelSource(yaml_dict["source_kernel_dir"])
+
+        for g in yaml_dict["list"]:
+            if "sysctl" in g:
+                self.kind = "sysctl"
+                group = g["sysctl"]
+                functions = g["functions"]
+            else:
+                group = None
+                functions = yaml_dict["list"]
+            self.fun_groups[group] = self.FunctionGroup()
+            for f in functions:
+                self.add_fun(f["name"],
+                             LlvmKernelModule(
+                                 os.path.join(os.path.relpath(
+                                     self.snapshot_source.kernel_dir),
+                                     f["llvm"]))
+                             if f["llvm"] else None,
+                             f["glob_var"],
+                             f["tag"],
+                             group)
+
+    def to_yaml(self):
+        """
+        Dump the snapshot as a YAML string. Paths to files are given relative
+        to the source snapshot directory.
+        :return: YAML string.
+        """
+        # Create a function list or a sysctl group list.
+        if None in self.fun_groups:
```

Comment:
True, however, IIRC it has been decided for the current version because the usage of a nested `for` itself is not that much readable in this case. Debatable, but I believe it would be better to leave this be for better clarity since the code improvement is relatively small (less than 10 lines and close to each other).

Reply:
I think the meaning of the nested list comprehension is clear in this case - one is for processing functions groups, the other is for processing functions.